pkg/middleware/log_server_error_request.go

```diff
@@ -0,0 +1,62 @@
+package middleware
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+
+	ce "github.com/content-services/content-sources-backend/pkg/errors"
+	"github.com/labstack/echo/v4"
+)
+
+const BodyDumpLimit = 1000
+const BodyStoreKey = "body_backup"
+
+func LogServerErrorRequest(next echo.HandlerFunc) echo.HandlerFunc {
+	return func(c echo.Context) (err error) {
+		if c.Get(BodyStoreKey) == nil {
+			storeRequestBody(c)
+		}
+		if err = next(c); err != nil {
+			if containsServerError(err) {
+				logRequestBody(c)
+			}
+			return err
+		}
+		return nil
+	}
+}
+
+func containsServerError(err error) bool {
+	httpError := new(ce.ErrorResponse)
+	if errors.As(err, httpError) {
+		for _, e := range httpError.Errors {
+			if e.Status >= http.StatusInternalServerError {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func logRequestBody(c echo.Context) {
+	if body := c.Get(BodyStoreKey); body != nil {
+		storedBodyBytes, ok := body.([]byte)
+		if !ok {
+			c.Logger().Error("Error reading request body")
+		}
+		c.Logger().Errorf("Request body: %v", string(storedBodyBytes))
+	}
+}
+
+func storeRequestBody(c echo.Context) {
+	var reqBody []byte
+	if c.Request().Body != nil {
+		reqBody, _ = io.ReadAll(c.Request().Body)
+	}
+	c.Request().Body = io.NopCloser(bytes.NewBuffer(reqBody))
+
+	limit := min(len(reqBody), BodyDumpLimit)
+	c.Set(BodyStoreKey, reqBody[:limit])
```

Comment:
I worry a bit about the performance hit of reading in the entire request body, especially for large requests coming in.

I played around a bit with bufio.Reader which implements a buffered Reader. This allows for 'peeking' at the first 'n' bytes without reading the entire Body. 

 This doesn't come with a Close() method though so you have to implement that:

```
type BufferedReadCloser struct {
	*bufio.Reader
	io.ReadCloser
}

func (rw BufferedReadWriteCloser) Close() error {
	return rw.ReadCloser.Close()
}

func (rw BufferedReadWriteCloser) Read(p []byte) (int, error) {
	return rw.Reader.Read(p)
}
```

Then you can create a buffered ReadCloser  from the request body:
```
		buffered := BufferedReadCloser{bufio.NewReader(c.Request().Body), c.Request().Body}
```

and 'peek' at the first 1000 bytes:

```
bytes, err := buffered.Peek(1000)
```

And then you could just set the buffered reader as the body:
```
c.Request().Body =  buffered
```

This way we aren't duplicating the entire body in memory, we're just buffering the first ~1000 bytes we read, at least until the request is actually processed.  
What do you think?

Reply:
True, this could be a more performant solution. 🤔 
And the `Peek()` method seems to **not** advance the reader so maybe I wouldn't have to restore the body on the request by writing the bytes read back to it. 😅
Will test it out. 👍🏼

Reply:
Did some profiling for this solution and it seems to be more memory efficient, but at the cost of CPU time. Although still an improvement I think, as the memory footprint might be a more important metric, so we should go with it 😄 
\
Methodology:
- 1,000,000 POST requests wrapped in the middleware each with body of ~12500 bytes.
- Code was an adjusted version of the `TestLargeBodyCutoff()` test case for this middleware.
- Profiled with the JetBrains (GoLand) built-in profiler.
- Averaged 3 runs.

Results:
|   | `ReadAll()` (Baseline)  | `Peek()` | Diff |
| -------------| ------------- | ------------- | ------------- |
| **Allocated Space**  | 97 GB | 74 GB | - 23.7 %  |
| **CPU Time** | 84 sec. | 104 sec. | + 23.8 % |

Reply:
interesting!!  Out of curiosity what does the cpu/memory look like without the middleware in use?

Reply:
found a mistake in my test case, had added some debug/check read and forgot it there, updated table here and it seems the buffered reader is a better solution all around 🆒😁 

updated results:
|                | W/O       | ReadAll() | Peek()    |
|----------------|-----------|-----------|-----------|
| Alocated Space | 32.8 GB   | 69.2 GB   | 37.2 GB   |
| CPU Time       | 52.3 sec. | 85.0 sec. | 68.3 sec. |

Reply:
wow! very cool, nice work getting real data!

Reply:
just to be sure, tested it in a more expected normal scenario, along with the regular more real like `echo` setup (ran the test for creating a repository 1M times with the regular payload, so not a crazy big payload that has to be read/binded after the middleware) 😄 
and if the reads after the middleware with the buffered reader wrapped request body, are any slower for those normal sized requests then it's not really noticeable, even at 1M requests, so I think we should be fine 😁

Reply:
[resolved in the latest commit](https://github.com/content-services/content-sources-backend/pull/773/commits/8a34ffb865d15367b1bdda71ef5420a7bcab6422), implemented the new buffered reader after finding that it provides considerable performance gains, good catch/proposal 👍🏼😄